neurolib/models/multimodel/builder/adex.py

```diff
@@ -0,0 +1,944 @@
+        """
+        exc_coupling = (
+            self.params["K_exc"] * coupling_variables["node_inh_exc"]
+            + self.params["c_global"] * self.params["K_exc_global"] * self.params["ext_rate"]
+        )
+        inh_coupling = self.params["K_inh"] * coupling_variables["node_inh_inh"]
+        exc_coupling_squared = (
+            self.params["K_exc"] * coupling_variables["node_inh_exc_sq"]
+            + self.params["c_global"] ** 2 * self.params["K_exc_global"] * self.params["ext_rate"]
+        )
+        inh_coupling_squared = self.params["K_inh"] * coupling_variables["node_inh_inh_sq"]
+
+        return (
+            exc_coupling,
+            inh_coupling,
+            exc_coupling_squared,
+            inh_coupling_squared,
+        )
+
+    def _derivatives(self, coupling_variables):
+        (I_mu, I_syn_mu_exc, I_syn_mu_inh, I_syn_sigma_exc, I_syn_sigma_inh, firing_rate,) = self._unwrap_state_vector()
+
+        exc_inp, inh_inp, exc_inp_sq, inh_inp_sq = self._compute_couplings(coupling_variables)
+
+        I_sigma = se.sqrt(
+            2
+            * self.params["J_exc_max"] ** 2
+            * I_syn_sigma_exc
+            * self.params["tau_syn_exc"]
+            * self.params["tau_m"]
+            / ((1.0 + exc_inp) * self.params["tau_m"] + self.params["tau_syn_exc"])
+            + 2
+            * self.params["J_inh_max"] ** 2
+            * I_syn_sigma_inh
+            * self.params["tau_syn_inh"]
+            * self.params["tau_m"]
+            / ((1.0 + inh_inp) * self.params["tau_m"] + self.params["tau_syn_inh"])
+            + self.params["sigma_ext"] ** 2
+        )
+
+        # get values from linear-nonlinear lookup table
+        firing_rate_now = self.callback_functions["firing_rate_lookup"](I_mu, I_sigma)
+        tau = self.callback_functions["tau_lookup"](I_mu, I_sigma)
+
+        d_I_mu = (
+            self.params["J_exc_max"] * I_syn_mu_exc
+            + self.params["J_inh_max"] * I_syn_mu_inh
+            + system_input(self.noise_input_idx[0])
+            + self.params["ext_current"]
+            - I_mu
+        ) / tau
+
+        d_I_syn_mu_exc = ((1.0 - I_syn_mu_exc) * exc_inp - I_syn_mu_exc) / self.params["tau_syn_exc"]
+
+        d_I_syn_mu_inh = ((1.0 - I_syn_mu_inh) * inh_inp - I_syn_mu_inh) / self.params["tau_syn_inh"]
+
+        d_I_syn_sigma_exc = (
+            (1.0 - I_syn_mu_exc) ** 2 * exc_inp_sq
+            + (exc_inp_sq - 2.0 * self.params["tau_syn_exc"] * (exc_inp + 1.0)) * I_syn_sigma_exc
+        ) / (self.params["tau_syn_exc"] ** 2)
+
+        d_I_syn_sigma_inh = (
+            (1.0 - I_syn_mu_inh) ** 2 * inh_inp_sq
+            + (inh_inp_sq - 2.0 * self.params["tau_syn_inh"] * (inh_inp + 1.0)) * I_syn_sigma_inh
+        ) / (self.params["tau_syn_inh"] ** 2)
+        # firing rate as dummy dynamical variable with infinitely fast
+        # fixed-point dynamics
+        d_firing_rate = -self.params["lambda"] * (firing_rate - firing_rate_now)
```

Comment:
`LAMBDA_SPEED = 10.0`, it's not "infinitely fast" is it?

Reply:
no, it's not. BUT with typical timescale as we model with (dt=0.1 or 0.01) it's more than enough.. with higher lambdas the model sometimes blow-up.. and with `lambda_speed = 10.0` the results are comparable as if not using this trick

Reply:
MHhh did you test this? What is the unit of this equation? ms? We can use this IMO but we should make sure that we don't introduce artifacts when using the AdEx like this... maybe compare bifurcation diagrams later some time. Anyways, I think it should be thoroughly tested before we recommend the use of this. What do you think?

Reply:
well, there is a test in `tests/multimodel/test_adex.py` that tests one node and one dummy small network between this implementation and original `neurolib`'s native implementation, meaning it will run both implementations and compare it with Pearson's `r`

but definitely agree; my original plan was after merging all the necessary bits and pieces to work on notebooks that'll thoroughly compare `neurolib`'s native models (ALN, Hopf, WC) with this MultiModel modular implementations. Another set of notebooks would be detailing how to actually use this to build any model of your liking

Reply:
Sound good! I think this is in an important point we need to keep in mind. The synaptic model of AdEx has time scales of 4 or so... so it's actually less than this one.